tensorflow_probability/python/stats/kendalls_tau.py

```diff
@@ -0,0 +1,166 @@
+    Tensor that contains the ordering of y values that are sorted.
+  """
+  exchanges = 0
+  num = tf.size(y)
+  k = tf.constant(1, tf.int32)
+  while tf.less(k, num):
+    for left in tf.range(0, num - k, 2 * k, dtype=tf.int32):
+      rght = left + k
+      rend = tf.minimum(rght + k, num)
+      tmp = tf.TensorArray(dtype=tf.int32, size=num)
+      m, i, j = 0, left, rght
+      while tf.less(i, rght) and tf.less(j, rend):
+        permij = aperm.gather([i, j])
+        yij = tf.gather(y, permij)
+        if tf.less_equal(yij[0], yij[1]):
+          tmp = tmp.write(m, permij[0])
+          i += 1
+        else:
+          tmp = tmp.write(m, permij[1])
+          # Explanation here
+          # https://www.geeksforgeeks.org/counting-inversions/.
+          exchanges += rght - i
+          j += 1
+        m += 1
+      while tf.less(i, rght):
+        tmp = tmp.write(m, aperm.read(i))
+        i += 1
+        m += 1
+      while tf.less(j, rend):
+        tmp = tmp.write(m, aperm.read(j))
+        j += 1
+        m += 1
+      aperm = aperm.scatter(tf.range(left, rend), tmp.gather(tf.range(0, m)))
+    k *= 2
+  return exchanges, aperm.stack()
+
+
+def kendalls_tau(y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
+  """Computes Kendall's Tau for two ordered lists.
+
+  Kendall's Tau measures the correlation between ordinal rankings. This
+  implementation is similar to the one used in scipy.stats.kendalltau.
+  Args:
+    y_true: A tensor that provides a true ordinal ranking of N items.
+    y_pred: A presumably model provided ordering of the same N items:
+
+  Returns:
+    Kendell's Tau, the 1945 tau-b formulation that ignores ordering of
+    ties, as a scalar Tensor.
+  """
+  y_true = tf.reshape(y_true, [-1])
+  y_pred = tf.reshape(y_pred, [-1])
+  y_pred.shape.assert_is_compatible_with(y_true.shape)
+  if tf.equal(tf.size(y_true), 0) or tf.equal(tf.size(y_pred), 0):
+    warnings.warn("y_true and y_pred tensors are not the same size.")
+    return np.nan
+  perm = tf.argsort(y_true)
+  n = tf.shape(perm)[0]
+  if tf.less(n, 2):
+    warnings.warn("Scalar tensors have no defined ordering.")
+    return np.nan
+
+  left = 0
+  # scan for ties, and for each range of ties do a argsort on
+  # the y_pred value. (TF has no lexicographical sorting, although
+  # jax can sort complex number lexicographically. Hmm.)
+  lexi = tf.TensorArray(tf.int32, size=n)
+  for i in tf.range(n):
+    lexi = lexi.write(i, perm[i])
+  for right in tf.range(1, n):
+    ytruelr = tf.gather(y_true, tf.gather(perm, [left, right]))
+    if tf.not_equal(ytruelr[0], ytruelr[1]):
+      sub = perm[left:right]
+      subperm = tf.argsort(tf.gather(y_pred, sub))
+      lexi = lexi.scatter(tf.range(left, right), tf.gather(sub, subperm))
+      left = right
+  sub = perm[left:n]
+  subperm = tf.argsort(tf.gather(y_pred, perm[left:n]))
+  lexi.scatter(tf.range(left, n), tf.gather(sub, subperm))
+
+  # This code follows roughly along with scipy/stats/stats.py v. 0.15.1
```

Comment:
Unfortunately this is somewhat problematic. Google has to be relatively pedantic about licensing, and although scipy's BSD license is 'compatible' with TFP's Apache2 license---so in principle one can distribute BSD-licensed code as part of an Apache 2 project---the [terms of the license](https://www.scipy.org/scipylib/license.html) require that we reproduce both the copyright of the Scipy authors and the full text of the BSD license in any derived work. That's not impossible, but it's a pain that would be nicer to avoid.

This code will in any case need some restructuring to use TF control flow ops (see comment above). I'd suggest trying to do that restructuring by thinking primarily about the math and algorithmic approaches that make sense in TFP, without explicitly referring to the scipy source code (even if the considerations are similar enough that they ultimately lead to similar solutions). That will help us avoid dealing with open-source license compatibility headaches.

Reply:
Indeed, there's only a procedural ordering of the steps in common, which are largely dictated by the algorithm for tau-b.
I've removed this reference and change the code to follow the notation laid forth in William Knight's 1966 paper which I also believe inspired the scipy implementation, although it's uncredited.